### PR TITLE
Improved params validation and main function.

### PR DESCRIPTION
**Pull request by filippolauria**
Added a lot of comments to it because it is a big improvement.

Fixed some errors in yandex scraper and get_soa function.
Rewritten se_result_process and also a test for it.

### CODE REVIEW

File: dnsrecon.py [hunk 1]
```diff
@@ -516,21 +516,26 @@ def se_result_process(res, found_hosts):
     an A and AAAA query for the IP of the found host. Prints and returns a dictionary
     with all the results found.
     """
-    returned_records = []
-    if found_hosts is None:
+    if not found_hosts:
         return None
+
+    returned_records = []
     for sd in found_hosts:
```
Comment (L1ghtn1ng) [inline dnsrecon.py:523]: please change this to something that is more meaning full

File: dnsrecon.py [hunk 2]
```diff
@@ -516,21 +516,26 @@ def se_result_process(res, found_hosts):
...
-        for sdip in res.get_ip(sd):
-            if re.search(r"^A|CNAME", sdip[0]):
-                print_status("\t {0} {1} {2}".format(sdip[0], sdip[1], sdip[2]))
-                if re.search(r"^A", sdip[0]):
-                    returned_records.extend([{"type": sdip[0], "name": sdip[1],
-                                              "address": sdip[2]}])
-                else:
-                    returned_records.extend([{"type": sdip[0], "name": sdip[1],
-                                              "target": sdip[2]}])
+        for type_, name_, address_or_target_ in res.get_ip(sd):
```
Comment (L1ghtn1ng) [inline dnsrecon.py:524]: please see sd comment above

File: dnsrecon.py [hunk 3]
```diff
@@ -1431,69 +1432,183 @@ def main():
...
+        print(f"DNSRecon version {__version__} ( http://www.darkoperator.com )")
+        sys.exit(0)
+
+    # ~ validating type param which is in the form: type1,type2,...,typeN
+    # ~ if the pattern is not correct or if there is an unknown type we exit
     type = arguments.type
+    types = []
+    if type:
+        type = type.lower().strip()
+        m = re.match(r'^([a-z]+,?)+$', type)
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1473]: change m to be regex_match

File: dnsrecon.py [hunk 4]
```diff
@@ -1431,69 +1432,183 @@ def main():
...
+        if not m:
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1474]: see above comment

File: dnsrecon.py [hunk 5]
```diff
@@ -1431,69 +1432,183 @@ def main():
...
+            print_error("This type of scan is not valid")
+            sys.exit(1)
+
+        valid_types = type_map.keys()
+        incorrect_types = [t for t in type.split(',') if t not in valid_types]
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1479]: change the t variable to something descriptive

File: dnsrecon.py [hunk 6]
```diff
@@ -1431,69 +1432,183 @@ def main():
...
+        if incorrect_types:
+            print_error("This type of scan is not in the list: {0}".format(','.join(incorrect_types)))
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1481]: Is this possible to have it as an f-string?

File: dnsrecon.py [hunk 7]
```diff
@@ -1431,69 +1432,183 @@ def main():
...
+        # ~ combining the types and the type_map, we obtain
+        # ~ dictionary_required which is a list of bool
+        # ~ where True means that a dictionary file is required
+        dictionary_required = [type_map[t]['dictionary'] for t in types]
+
+    # ~ if a dictionary file is required,
+    # ~ we check that user has provided a valid one
+    if any(dictionary_required):
+        if arguments.dictionary:
+            d = arguments.dictionary.strip()
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1570]: No single char as a vairable

File: dnsrecon.py [hunk 8]
```diff
@@ -1431,69 +1432,183 @@ def main():
...
+            # ~ if the dictionary file provided by the user is valid,
+            # ~ we use it
+            if os.path.isfile(d):
+                dictionary = d
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1574]: see comment above

File: dnsrecon.py [hunk 9]
```diff
@@ -1431,69 +1432,183 @@ def main():
...
+            # ~ otherwise we try to load a default dictionary file
+            print_status('No dictionary file has been specified.')
+            script_dir = os.path.dirname(os.path.realpath(__file__)) + os.sep
+            dictionaries = ['/etc/dnsrecon/namelist.txt', script_dir + 'namelist.txt']
+
+            dictionary = ""
+            for d in dictionaries:
+                if os.path.isfile(d):
+                    dictionary = d
+                    break
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1588]: So single char as a variable

File: dnsrecon.py [hunk 10]
```diff
@@ -1431,69 +1432,183 @@ def main():
...
+
+            if not dictionary:
+                # ~ if we cannot find a valid dictionary file,
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1591]: Can you please remove ~ from all comments and just have a space instead

File: dnsrecon.py [hunk 11]
```diff
@@ -1431,69 +1432,183 @@ def main():
...
+                # ~ we exit
+                print_error("Cannot find the dictionary file provided with the tool.")
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1593]: Please use single quotes ever where it is applicable to use them to keep consistency with the quotes

File: dnsrecon.py [hunk 12]
```diff
@@ -1431,69 +1432,183 @@ def main():
...
-    if arguments.range:
-        ip_list = process_range(arguments.range)
-        if len(ip_list) > 0:
-            if type is None:
-                type = "rvl"
-            elif not re.search(r"rvl", type):
-                type = "rvl," + type
-        else:
-            print_error('Failed CIDR or Range is Required for type rvl')
+        print_status(f"Using the dictionary file: {dictionary}")
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1596]: see above comment

File: dnsrecon.py [hunk 13]
```diff
@@ -1507,214 +1622,144 @@ def main():
...
     # Set the resolver
     res = DnsHelper(domain, ns_server, request_timeout, proto)
 
-    domain_req = ['axfr', 'std', 'srv', 'tld', 'bing', 'yand', 'crt', 'zonewalk']
     scan_info = [" ".join(sys.argv), str(datetime.datetime.now())]
 
-    if type is not None:
+    # ~ we have finished to validate params,
+    # ~ we can start the execution
+    for t in types:
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1633]: no single char vars

File: dnsrecon.py [hunk 14]
```diff
@@ -1507,214 +1622,144 @@ def main():
...
 
-        # Check for any illegal enumeration types from the user
-        valid_types = ['axfr', 'std', 'rvl', 'brt', 'srv', 'tld', 'bing', 'yand', 'crt', 'snoop', 'zonewalk']
-        incorrect_types = [t for t in type.split(',') if t not in valid_types]
-        if incorrect_types:
-            print_error("This type of scan is not in the list: {0}".format(','.join(incorrect_types)))
+        # ~ we check if this type of scan requires the domain
+        if type_map[t]['domain'] and not domain:
+            print_error(f"{t}: No Domain to target specified!")
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1637]: see above comment

File: dnsrecon.py [hunk 15]
```diff
@@ -1507,214 +1622,144 @@ def main():
...
-                            json_file is not None):
-                        returned_records.extend(ds_zone_walk(res, domain, request_timeout))
-                    else:
-                        ds_zone_walk(res, domain, request_timeout)
-
-                else:
-                    print_error(f"This type of scan is not in the list: {r}")
+        try:
+            # ~ here we start checking for the different types
+            if t == 'axfr':
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1642]: see above comment

File: dnsrecon.py [hunk 16]
```diff
@@ -1507,214 +1622,144 @@ def main():
...
-                print_error(f"Could not resolve domain: {domain}")
-                sys.exit(1)
+                returned_records.extend(zonercds)
 
-            except dns.exception.Timeout:
-                print_error("A timeout error occurred please make sure you can reach the target DNS Servers")
-                print_error(f"directly and requests are not being filtered. Increase the timeout from {request_timeout} seconds")
-                print_error("to a higher number with --lifetime <time> option.")
-                sys.exit(1)
+            elif t == 'std':
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1651]: see above comment about no single chars

File: dnsrecon.py [hunk 17]
```diff
@@ -1507,214 +1622,144 @@ def main():
...
+                    returned_records.extend(std_enum_records)
 
-        # if an output xml file is specified it will write returned results.
-        if output_file is not None:
-            print_status(f"Saving records to XML file: {output_file}")
-            xml_enum_doc = dns_record_from_dict(returned_records, scan_info, domain)
-            write_to_file(xml_enum_doc, output_file)
+            elif t == 'rvl':
+                if not rvl_ip_list:
+                    print_error(f"{t}: Invalid Address/CIDR or Address Range provided.")
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1661]: see above

File: dnsrecon.py [hunk 18]
```diff
@@ -1507,214 +1622,144 @@ def main():
...
+                    continue
 
-        # if an output db file is specified it will write returned results.
-        elif results_db is not None:
-            print_status(f"Saving records to SQLite3 file: {results_db}")
-            create_db(results_db)
-            write_db(results_db, returned_records)
+                # ~ print_status(f"{t}: Performing Reverse Look-up for {rvl_ip_list}...")
+                rvl_enum_records = brute_reverse(res, rvl_ip_list, verbose,
+                                                 thread_num=thread_num)
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1666]: this can be all on 1 line

File: dnsrecon.py [hunk 19]
```diff
@@ -1507,214 +1622,144 @@ def main():
...
+            elif t == "crt":
+                print_status(f"{t}: Performing Crt.sh Search Enumeration against {domain}...")
+                crt_enum_records = se_result_process(res, scrape_crtsh(domain))
+                if do_output:
+                    returned_records.extend(crt_enum_records)
+
+            elif t == "snoop":
+                if not (dictionary and ns_server):
+                    print_error(f"{t}: A dictionary file and at least one Name Server have to be specified!")
+                    continue
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1712]: no single var chars

File: dnsrecon.py [hunk 20]
```diff
@@ -1507,214 +1622,144 @@ def main():
...
+                print_status(f"{t}: Performing Cache Snooping against NS Server: {ns_server[0]}...")
+                cache_enum_records = in_cache(res, dictionary, ns_server[0])
+                if do_output:
+                    returned_records.extend(cache_enum_records)
 
-        # if an output json file is specified it will write returned results.
-        elif json_file is not None:
-            print_status(f"Saving records to JSON file: {json_file}")
-            write_json(json_file, returned_records, scan_info)
+            elif t == "zonewalk":
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1719]: see above

File: dnsrecon.py [hunk 21]
```diff
@@ -1507,214 +1622,144 @@ def main():
...
+            elif t == "snoop":
+                if not (dictionary and ns_server):
+                    print_error(f"{t}: A dictionary file and at least one Name Server have to be specified!")
+                    continue
 
-        # if an output csv file is specified it will write returned results.
-        elif csv_file is not None:
-            print_status(f"Saving records to CSV file: {csv_file}")
-            write_to_file(make_csv(returned_records), csv_file)
+                print_status(f"{t}: Performing Cache Snooping against NS Server: {ns_server[0]}...")
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1714]: see above

File: dnsrecon.py [hunk 22]
```diff
@@ -1507,214 +1622,144 @@ def main():
...
-            print_status(f"Saving records to JSON file: {json_file}")
-            write_json(json_file, returned_records, scan_info)
+            elif t == "zonewalk":
+                zonewalk_result = ds_zone_walk(res, domain, request_timeout)
+                if do_output:
+                    returned_records.extend(zonewalk_result)
 
-        sys.exit(0)
+            else:
+                print_error(f"{t}: This type of scan is not in the list.")
```
Comment (L1ghtn1ng) [inline dnsrecon.py:1725]: see above

File: lib/yandexenum.py [hunk 23]
```diff
@@ -42,19 +42,19 @@ def scrape_yandex(dom):
...
         try:
-            sock = urllib.request.urlopen(url)
-            data = sock.read()
-        except AttributeError:
             sock = urllib.request.urlopen(url)
             data = sock.read().decode("utf-8")
+            sock.close()
+        except Exception as e:
+            print_error(e)
+            return []
```
Comment (L1ghtn1ng) [inline lib/yandexenum.py:50]: is this really needed?

**Review by L1ghtn1ng — CHANGES_REQUESTED** [review]
please see comments that have been left

**Comment by L1ghtn1ng** [comment]
Looking good with the new changes, still some changes need doing though

**Comment by L1ghtn1ng** [comment]
@filippolauria Any update on when you will have finished the changes?

**Comment by filippolauria** [comment]
Hello, most of the requested changes have been made. Please start merging them. They already solve most of the bugs and issues with the program, other than provide different optimizations ti the code.

**Pull request merged**